ankimorphs/jit_highlight_morphs.py

```diff
@@ -0,0 +1,138 @@
+from __future__ import annotations
+
+import re
+
+from anki.template import TemplateRenderContext
+
+from .ankimorphs_config import (
+    AnkiMorphsConfig,
+    AnkiMorphsConfigFilter,
+    get_matching_filter,
+)
+from .ankimorphs_db import AnkiMorphsDB
+from .ankimorphs_globals import EXTRA_FIELD_HIGHLIGHTED
+from .morpheme import Morpheme
+from .morphemizers import spacy_wrapper
+from .morphemizers.morphemizer import (
+    Morphemizer,
+    SpacyMorphemizer,
+    get_morphemizer_by_description,
+)
+from .text_highlighting import get_highlighted_text
+from .text_preprocessing import (
+    get_processed_morphemizer_morphs,
+    get_processed_spacy_morphs,
+)
+
+
+def am_highlight_morphs(
+    field_text: str,
+    field_name: str,
+    filter_name: str,
+    context: TemplateRenderContext,
+) -> str:
+    """Use morph learning progress to decorate the morphemes in the supplied text.
+    Adds css classes to the output that can be styled in the card."""
+
+    if filter_name != "am-highlight" or field_name == EXTRA_FIELD_HIGHLIGHTED:
+        return field_text
+
+    am_config_filter: AnkiMorphsConfigFilter | None = get_matching_filter(
+        context.note()
+    )
+
+    if am_config_filter is None:
+        return field_text
+
+    morphemizer: Morphemizer | None = get_morphemizer_by_description(
+        am_config_filter.morphemizer_description
+    )
+
+    if not morphemizer:
+        return field_text
+
+    am_config = AnkiMorphsConfig()
+
+    card_morphs: list[Morpheme] = get_morphemes(morphemizer, am_config, field_text)
+
+    if not card_morphs:
+        return field_text
+
+    return get_highlighted_text(am_config, card_morphs, field_text)
+
+
+def get_morphemes(
+    morphemizer: Morphemizer,
+    am_config: AnkiMorphsConfig,
+    field_text: str,
+) -> list[Morpheme]:
+    """Take in a string and gather the morphemes from it."""
+
+    # If we were piped in after the `furigana` built-in filter, or if there is html in the source
+    # data, we need to do some unpacking.
+    #
+    clean_field_text = dehtml(field_text)
+
+    if isinstance(morphemizer, SpacyMorphemizer):
+        nlp = spacy_wrapper.get_nlp(
+            morphemizer.get_description().removeprefix("spaCy: ")
+        )
+
+        all_morphs = get_processed_spacy_morphs(
+            am_config, next(nlp.pipe([clean_field_text]))
+        )
+    else:
+        all_morphs = get_processed_morphemizer_morphs(
+            morphemizer, clean_field_text, am_config
+        )
+
+    return get_morph_stats(list(set(all_morphs)), am_config)
+
+
+def get_morph_stats(
+    morphs: list[Morpheme], am_config: AnkiMorphsConfig
+) -> list[Morpheme]:
+    if not morphs:
+        return []
+
+    am_db = AnkiMorphsDB()
```

Comment:
The db connection needs to be closed. If we have a short scope then using a context manager takes care of it:

https://github.com/mortii/anki-morphs/blob/cf3f930c9f0dddc44a0b410a7eed7a60ba1e2400/ankimorphs/__init__.py#L166

If we want a longer lived/scoped object for performance reason, or it's too nested, then we can manually call `am_db.con.close()` instead.

Reply:
awesome! I'll wrap it in a `with`.